### PR TITLE
Np 116 - implementing call to EIS

### PR DESCRIPTION
**Pull request by schemabotview**
implementing call to EIS

### CODE REVIEW

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/CustomHeaderNames.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.uknwauthcheckerapi.models
+
+object CustomHeaderNames {
+  val environment    = "Environment"
+  val correlationId  = "CorrelationId"
+  val xApiKey        = "X-API-Key"
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/CustomHeaderNames.scala]: What is this?

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import play.api.http.Status._
+import play.api.libs.json._
+
+sealed abstract class ErrorCode(val code: String, val statusCode: Int) extends Product with Serializable
+
+object ErrorCode {
+  case object BadRequest extends ErrorCode("BAD_REQUEST", BAD_REQUEST)
+
+  case object BadGateway extends ErrorCode("BAD_GATEWAY", BAD_GATEWAY)
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: This is not a supported statusf from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  case object NotFound extends ErrorCode("NOT_FOUND", NOT_FOUND)
+
+  case object Forbidden extends ErrorCode("FORBIDDEN", FORBIDDEN)
+
+  case object InternalServerError extends ErrorCode("INTERNAL_SERVER_ERROR", INTERNAL_SERVER_ERROR)
+
+  case object GatewayTimeout extends ErrorCode("GATEWAY_TIMEOUT", GATEWAY_TIMEOUT)
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  case object UnsupportedMediaType extends ErrorCode("UNSUPPORTED_MEDIA_TYPE", UNSUPPORTED_MEDIA_TYPE)
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+              s"""
+                   |Message: ${ex.getMessage}
+                   |""".stripMargin
+            }
+            .getOrElse("No exception is available")
+          logger.error(s"""Internal Server Error: $message
+               |
+               |$causeText""".stripMargin)
+          Future.successful(())
+        case BadGateway(message, _, responseCode)    =>
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      }
+  }
+
+  trait Converter[E] {
+    def convert(error: E): ResponseError
+  }
+
+  implicit val IntegrationFrameworkErrorConverter: Converter[IntegrationFrameworkError] =
+    {
+      case IntegrationFrameworkError.BadGateway(cause, statusCode) => ResponseError.badGateway(cause, statusCode)
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/IntegrationFrameworkError.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.uknwauthcheckerapi.models
+
+sealed trait IntegrationFrameworkError
+
+object IntegrationFrameworkError {
+  case class InternalUnexpectedError(cause: Option[Throwable]) extends IntegrationFrameworkError
+  case class BadGateway(reason: String, code: Int) extends IntegrationFrameworkError
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/IntegrationFrameworkError.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ResponseError.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  def badRequestError(message: String): ResponseError =
+    StandardError(message, ErrorCode.BadRequest)
+
+  def unauthorisedError(message: String): ResponseError =
+    StandardError(message, ErrorCode.Unauthorized)
+
+  def notFoundError(message: String): ResponseError =
+    StandardError(message, ErrorCode.NotFound)
+
+  def badGateway(message: String, code: Int): ResponseError =
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ResponseError.scala]: This is not a supported status from EIS/PDS

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ResponseError.scala [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      (__ \ messageFieldName).read[String] and
+        (__ \ codeFieldName).read[ErrorCode]
+    )(StandardError.apply _)
+
+  def unapply(error: ResponseError): Option[(String, ErrorCode)] = Some((error.message, error.code))
+}
+
+case class StandardError(message: String, code: ErrorCode) extends ResponseError
+
+case class BadGateway(
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/models/ResponseError.scala]: This is not a supported status from EIS/PDS

File: conf/application.conf [hunk 10]
```diff
@@ -19,8 +19,14 @@ include "backend.conf"
 
 appName = uknw-auth-checker-api
 
+authType = "EIR"
```
Comment (TheDanielDoyle) [inline conf/application.conf]: This should be UKNW

File: resources/public/api/conf/1.0/application.yaml [hunk 11]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (TheDanielDoyle) [inline resources/public/api/conf/1.0/application.yaml]: This space shouldn't exist

File: test-common/uk/gov/hmrc/uknwauthcheckerapi/generators/Generators.scala [hunk 12]
```diff
@@ -18,24 +18,28 @@ package uk.gov.hmrc.uknwauthcheckerapi.generators
...
+      eoris <- eorisGen
+    } yield AuthorisationRequest(date, eoris)
+  }
+
+  implicit val arbEisAuthorisationRequest: Arbitrary[EisAuthorisationRequest] = Arbitrary {
+    for {
+      date  <- Gen.option(LocalDate.now())
       eoris <- eorisGen
-    } yield AuthorisationRequest(date = date, eoris = eoris)
+    } yield EisAuthorisationRequest(date, "EIR", eoris)
```
Comment (TheDanielDoyle) [inline test-common/uk/gov/hmrc/uknwauthcheckerapi/generators/Generators.scala]: This should be UKNW

File: test-common/uk/gov/hmrc/uknwauthcheckerapi/generators/TestData.scala [hunk 13]
```diff
@@ -16,13 +16,35 @@
...
 
   def randomAuthorisationRequest: AuthorisationRequest = arbAuthorisationRequest.arbitrary.sample.get
+
+  implicit val arbValidGetAuthorisationsResponse: Arbitrary[ValidGetAuthorisationsResponse] = Arbitrary {
+    for {
+      date  <- Gen.option(LocalDate.now())
+      eoris <- eorisGen
+    } yield ValidGetAuthorisationsResponse(
+      EisAuthorisationsResponse(
+        date.getOrElse(LocalDate.now()), "EIR", eoris.map(e => {
```
Comment (TheDanielDoyle) [inline test-common/uk/gov/hmrc/uknwauthcheckerapi/generators/TestData.scala]: This should be UKNW

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/CustomHeaderNames.scala (hunk 1 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/CustomHeaderNames.scala]: Removed unwanted headers.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala (hunk 2 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala (hunk 3 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala (hunk 4 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorCode.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala (hunk 5 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala (hunk 6 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/ErrorHandler.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/models/IntegrationFrameworkError.scala (hunk 7 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/models/IntegrationFrameworkError.scala]: Removed stale classes.

File: app/uk/gov/hmrc/uknwauthcheckerapi/connectors/BaseConnector.scala [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        .recoverTotal(error => Future.failed(JsResult.Exception(error)))
+  }
+
+  implicit class RequestBuilderHelpers(requestBuilder: RequestBuilder) {
+    def executeAndDeserialise[T](implicit ec: ExecutionContext, reads: Reads[T]): Future[T] =
+      requestBuilder
+        .execute[HttpResponse]
+        .flatMap { response =>
+          response.status match {
+            case OK | CREATED | ACCEPTED => response.as[T]
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/connectors/BaseConnector.scala]: Ok is the only acceptable response here

File: app/uk/gov/hmrc/uknwauthcheckerapi/connectors/IntegrationFrameworkConnector.scala [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+      (HeaderNames.ACCEPT, "application/json"),
+      (HeaderNames.AUTHORIZATION, s"Bearer $bearerToken")
+    )
+  }
+
+  def getEisAuthorisationsResponse(eisAuthorisationRequest: EisAuthorisationRequest)
+                                  (implicit hc: HeaderCarrier): Future[EisAuthorisationsResponse] =
+    retryFor[EisAuthorisationsResponse]("Integration framework Response")(retryCondition) {
+      httpClient
+        .post(url"${appConfig.baseUrl("integration-framework")}/authorisations")
```
Comment (TheDanielDoyle) [inline app/uk/gov/hmrc/uknwauthcheckerapi/connectors/IntegrationFrameworkConnector.scala]: This is the wrong endpoint. See chat

**Review by TheDanielDoyle — CHANGES_REQUESTED** [review]

File: conf/application.conf (hunk 10 above)
Comment (schemabotview) [reply conf/application.conf]: changed to UKNW

File: resources/public/api/conf/1.0/application.yaml (hunk 11 above)
Comment (schemabotview) [reply resources/public/api/conf/1.0/application.yaml]: Removed space at line 1

File: app/uk/gov/hmrc/uknwauthcheckerapi/connectors/BaseConnector.scala (hunk 14 above)
Comment (schemabotview) [reply app/uk/gov/hmrc/uknwauthcheckerapi/connectors/BaseConnector.scala]: Retained only OK response status code.

**Review by TheDanielDoyle — APPROVED** [review]

**Pull request merged**